docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,383 @@
+following changes:
+
+* get holdings: query for holdings owned by the public key, for all official
+token programs listed in the registry (not just `Tokenkeg...`)
+* get balances: avoid deserializing holding data, and instead use
+`getBalance` from RPC
+* transfer tokens: only use `ApproveChecked` and `TransferChecked` on the token
+program
+* create holding: avoid directly creating an account and initializing,
+and use the associated token account program instead
+
+### On-chain Programs
+
+To properly support i-tokens on-chain, programs will need to make these changes:
+
+* multiple token programs: all programs that support SPL Tokens must always
+include the program id of the token(s) holdings they're passing in.
+For example, token-swap needs to accept separate token programs for token A and B
+* hard-coded token program: avoid using `Tokenkeg...` directly, and delete `id()`
+the SPL token code
+* get balance: instead of deserializing the account data into an SPL holding, use
+the new wrapper to deserialize or call `GetHoldingBalance` instruction with the
+appropriate token program
+* get owner: same as above, but using the `GetHoldingOwner`
+* get supply: same as above, but using the `GetMintSupply`
+* get mint authority: same as above, but using the `GetMintAuthority`
+* get mint decimals: same as above, but using the `GetMintDecimals`
+* transfer: only use `TransferChecked`, which requires passing mints
+
+## Other New Token Programs
+
+For awhile now, people have been coming up with ideas that likely require a new
+token program to implement. These include:
+
+### Voting Tokens
+
+Outlined in a [GitHub issue](https://github.com/solana-labs/solana-program-library/issues/131),
+voting tokens use additional data about how long a token has been held.
+
+### Non-Fungible Tokens
+
+It will be possible to add a lot of metadata regarding owners and traders, either
+on a new mint or NFT account.
+
+### Allow-list Tokens
+
+Only a certain list keys are allowed (or not) to transfer to tokens, controlled
+by the mint.
+
+## Terminology
+
+* token amount: the amount of tokens as seen from the outside. For example, an
+interest-bearing token holding with 10 in token amount is converted to a proportional
+ownership of 0.5% of the total supply. This amount is always shown in UIs, and
+also cached in the `amount` field of 
+
+* share amount: the proportional ownership of a holding in terms of the total
+supply of shares. Internally, the interest-bearing token program uses shares to
+mint / transfer / burn, and this amount is never shown in UIs.
+
+## Proposed Layout
+
+### Mint
+
+```rust
+pub struct Mint {
+    // SPL Mint Fields
+    pub mint_authority: COption<Pubkey>,
+    pub supply: u64, // Cached token amount of current supply, used by RPC
+    pub decimals: u8,
+    pub is_initialized: bool,
+    pub freeze_authority: COption<Pubkey>,
+
+    // Interest-bearing Mint Fields
+    pub share_supply: f64, // The total amount of shares outstanding
```

Comment:
Hopefully we won't actually use f64 for anything, though ;)

Reply:
From a conversation with Jack, who was talking about how well they work in the vm :-)